src/test/java/io/debezium/connector/vitess/AbstractVitessConnectorTest.java

```diff
@@ -200,8 +200,11 @@ protected static String incrementGtid(String gtid, int increment) {
         return gtid.substring(0, idx) + seq;
     }
 
-    protected TestConsumer testConsumer(int expectedRecordsCount, String... topicPrefixes) {
-        return new TestConsumer(expectedRecordsCount, topicPrefixes);
+    protected TestConsumer testConsumer(int expectedRecordsCount, String... topicPrefixes) throws InterruptedException {
+        TestConsumer consumer = new TestConsumer(expectedRecordsCount, topicPrefixes);
+        // Sleep for 1 second before return to avoid race conditions.
+        Thread.sleep(1000);
```

Comment:
Can we avoid such unconditional thread sleep call? It's not reliable typically and it increases execution time of the test suite. What's the race condition about? Typically, the better fix is to use Awaitility to make sure some specific condition has been matched in a test.

Reply:
+1

could you also comment a bit around where does the race condition come from? would [stopConnector](https://github.com/debezium/debezium/blob/main/debezium-embedded/src/test/java/io/debezium/embedded/AbstractConnectorTest.java#L146) help to handle that?

Reply:
Others have chimed in, but just wanted to add a thought. The general things you could look out for with sleeps is (1) making unit tests longer than they need to and (2) more importantly, sleeps having potential to make unit tests flaky.

Depending on the code you could do something with Awaitility as Gunnar said like `await().until(<something>)` or some other variant of the API. Another thing is to inject a Clock so we can control/advance them as we need in our tests -- although this could be a larger refactor, for another time (I haven't looked as deeply into this part of the codebase yet).

Reply:
All the Vitess connector's IT tests always failed on my laptop due to similar reason to https://issues.redhat.com/browse/DBZ-4391 . Because the expected data change events could be observed

After troubleshooting for a while, I think there is a race condition. A typical example would be starting from this line https://github.com/debezium/debezium-connector-vitess/blob/main/src/test/java/io/debezium/connector/vitess/VitessConnectorIT.java#L116

When the VStream is started it used "current" as starting gtid position. The test inserts seems started too early when the connector started VStream, so when VTGate receives the connection the insert already completed; this makes VTGate only emit new changes after the insert. This means the insert change event is "lost" for us in the test cases and expecting the record would time out.

Reply:
If set the GTID to the last GTID in the current Vitess mysql, by passing in configBuilder in the tests, like this:
```
        builder = builder
                .with(RelationalDatabaseConnectorConfig.SERVER_NAME, TEST_SERVER)
                .with(VitessConnectorConfig.VTGATE_HOST, VTGATE_HOST)
                .with(VitessConnectorConfig.VTGATE_PORT, VTGATE_PORT)
                .with(VitessConnectorConfig.VTGATE_USER, USERNAME)
                .with(VitessConnectorConfig.VTGATE_PASSWORD, PASSWORD)
                .with(VitessConnectorConfig.GTID, "MySQL56/0588557d-5357-11ec-84d5-0242ac110002:1-1264")
                .with(VitessConnectorConfig.POLL_INTERVAL_MS, 100);
```
The tests will work. This means it's defintely due to the race condition of VStream with "current"  and inserts as mentioned above.

Reply:
I see, that makes sense. I was using [a slightly different approach](https://github.com/debezium/debezium-connector-vitess/blob/e6aebf00602189d125c88002b5b576b167208732/src/test/java/io/debezium/connector/vitess/VitessReplicationConnectionIT.java#L77-L83) to tackle this exact problem - it was taking advantage that VStream will pass along an event with "OTHER" type when there is no writes happening.

But I think the logging approach you have is more reliable, could you also change it in `VitessReplicationConnectionIT` as well so that we are consistent on that front?